kernel/src/snapshot.rs

```diff
@@ -67,6 +76,167 @@ impl Snapshot {
+        let fs_client = engine.get_file_system_client();
+
+        // start from either the checkpoint version + 1 or version 1
+        let start_version =
+            if let Some(checkpoint_version) = existing_snapshot.log_segment.checkpoint_version {
+                checkpoint_version + 1
+            } else {
+                // if there is no checkpoint, the existing log segment must start at 0, thus, we
+                // list from 1
+                1
+            };
+
+        // Check for new commits
+        let (new_ascending_commit_files, checkpoint_parts) =
+            log_segment::list_log_files_with_version(
+                fs_client.as_ref(),
+                &log_root,
+                Some(start_version),
+                new_version,
+            )?;
+
+        // NB: we need to check both checkpoints and commits since we filter commits at and below
+        // the checkpoint version. Example: if we have a checkpoint + commit at version 1, the log
+        // listing above will only return the checkpoint and not the commit.
+        if new_ascending_commit_files.is_empty() && checkpoint_parts.is_empty() {
+            match new_version {
+                Some(new_version) if new_version != old_version => {
+                    // No new commits, but we are looking for a new version
+                    return Err(Error::Generic(format!(
+                        "Requested snapshot version {} is newer than the latest version {}",
+                        new_version, old_version
+                    )));
+                }
+                _ => {
+                    // No new commits, just return the same snapshot
+                    return Ok(existing_snapshot.clone());
+                }
+            }
+        }
+
+        // create a log segment just from existing_checkpoint.version -> new_version
+        // OR could be from 1 -> new_version
+        let mut new_log_segment = LogSegment::try_new(
+            new_ascending_commit_files,
+            checkpoint_parts,
+            log_root.clone(),
+            new_version,
+        )?;
+
+        let new_end_version = new_log_segment.end_version;
+        match new_end_version.cmp(&old_version) {
+            std::cmp::Ordering::Less => {
+                // we should never see a new log segment with a version < the existing snapshot
+                // version, that would mean a commit was incorrectly deleted from the log
+                return Err(Error::Generic(format!(
+                "Unexpected state: The newest version in the log {} is older than the old version {}",
+                new_end_version, old_version
+            )));
+            }
+            std::cmp::Ordering::Equal => {
+                // No new commits, just return the same snapshot
+                return Ok(existing_snapshot.clone());
+            }
+            std::cmp::Ordering::Greater => (), // expected
+        }
+
+        if !new_log_segment.checkpoint_parts.is_empty() {
```

Comment:
easier to read like this?
```suggestion
        if new_log_segment.checkpoint_version.is_some() {
```